tests/charms/tls_certificates_interface/v1/test_tls_certificates_v1_provides.py

```diff
@@ -0,0 +1,477 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import json
+import unittest
+from unittest.mock import PropertyMock, call, patch
+
+from ops import testing
+
+from tests.charms.tls_certificates_interface.v1.dummy_provider_charm.src.charm import (
+    DummyTLSCertificatesProviderCharm,
+)
+
+testing.SIMULATE_CAN_CONNECT = True
+
+BASE_CHARM_DIR = "tests.charms.tls_certificates_interface.v1.dummy_provider_charm.src.charm.DummyTLSCertificatesProviderCharm"  # noqa: E501
+LIB_DIR = "lib.charms.tls_certificates_interface.v1.tls_certificates"
+
+
+def _load_relation_data(raw_relation_data: dict) -> dict:
+    """Loads relation data from the relation data bag.
+
+    Json loads all data.
+
+    Args:
+        raw_relation_data: Relation data from the databag
+
+    Returns:
+        dict: Relation data in dict format.
+    """
+    certificate_data = dict()
+    for key in raw_relation_data:
+        try:
+            certificate_data[key] = json.loads(raw_relation_data[key])
+        except json.decoder.JSONDecodeError:
+            certificate_data[key] = raw_relation_data[key]
+    return certificate_data
+
+
+class Test(unittest.TestCase):
+    def setUp(self):
+        self.relation_name = "certificates"
+        self.harness = testing.Harness(DummyTLSCertificatesProviderCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+        self.maxDiff = None
+
+    @patch(
+        f"{LIB_DIR}.CertificatesProviderCharmEvents.certificate_creation_request",
+        new_callable=PropertyMock,
+    )
+    def test_given_csr_in_relation_data_when_relation_changed_then_certificate_creation_request_is_emitted(  # noqa: E501
+        self, patch_certificate_creation_request
+    ):
+        remote_app = "tls-certificates-requirer"
+        remote_unit_name = "tls-certificates-requirer/0"
+        relation_id = self.harness.add_relation(
+            relation_name=self.relation_name, remote_app=remote_app
+        )
+        self.harness.add_relation_unit(relation_id=relation_id, remote_unit_name=remote_unit_name)
```

Comment:
How about adding relation in `setUp` so that you don't have to do it in every test?

Reply:
I created a method to cover relation creation. I haven't moved them to set up because I think the test author needs to be aware they are creating relations.

Reply:
Works for me.